### ISSUE TITLE
[VL] Gluten is Slower than Vanilla Spark

### ISSUE THREAD

**Issue by ted-jenks** [oap-project/gluten#3380]
### Backend

VL (Velox)

### Bug description

We have been testing Gluten on our production setup to see if it could be useful. On a TPC-DS benchmark at scale 1000, we see gluten is faster for only 32 of the 100 total queries. Overall, the performance isn’t good enough to justify using it. 

I was wondering if there is anything we could be doing with our configuration or if there are other similar things that can improve the performance? For instance, are there particular architectures which we are likely to see more sizable performance improvements for?

![benchmark_results_min_3](https://github.com/oap-project/gluten/assets/77486246/4069fedb-6b97-4a55-94dc-16e128da020a)

### Spark version

Spark-3.2.x

### Spark configurations

```
spark.plugins: io.glutenproject.GlutenPlugin
spark.shuffle.manager: org.apache.spark.shuffle.sort.ColumnarShuffleManager
spark.gluten.sql.columnar.backend.lib: velox
spark.memory.offHeap.enabled: true
spark.memory.offHeap.size: 30g
spark.executor.extraJavaOptions: --illegal-access=permit -Dio.netty.tryReflectionSetAccessible=true --add-opens java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED
spark.driver.extraJavaOptions: --illegal-access=permit -Dio.netty.tryReflectionSetAccessible=true --add-opens java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED
spark.gluten.sql.columnar.shuffle.writeEOS: false
spark.gluten.loadLibFromJar: true
spark.driver.memory: 13g
spark.driver.maxResultSize: 8g
spark.executor.memory: 27g
spark.executor.memoryOverhead: 3g
spark.executor.instances: 16
spark.dynamicAllocation.maxExecutors: 16
```

### System information

We are running Kubernetes on cloud AMD64 instances.

**Comment by zhanglistar** [comment]
What's the IO metrics when running SQLs?

**Comment by kelvin-qin** [comment]
The physical environment has a significant impact on performance, such as the network. Disk read/write speeds, cpu model, etc. You need to add these.

**Comment by kelvin-qin** [comment]
I'm not sure how it performs in a k8s environment, but in general cloud environments are worse. My own test results proved that at least 70% of SQL achieved performance gains. It's not unusual for a few SQLs to be slower than spark, but your test results are still very different from mine.

**Comment by kelvin-qin** [comment]
I use 3 x86 physical machines with 12 HDDs and 10Gigabit networks.

**Comment by ted-jenks** [comment]
> I'm not sure how it performs in a k8s environment, but in general cloud environments are worse.

Definitely, but this is a like-for-like comparison (in fact, it was slightly beneficial to gluten as we allocated it the 30g off-heap). We ran each the benchmark three times and these are the results with a variance of maybe +/-0.1 of the speedup results.

I find it surprising as the conclusion from these data is something like "native execution is slower than java"; clearly that is not the case, so there must be something our set up holding gluten up. Another thing that may be relevant: we are building from source (off main branch) using the default build set up.

We are not seeing anything like executor loss or higher stage retry rates in the gluten runs. In the slower queries, stages just run slower than the equivalent stages in vanilla spark. This is not isolated to IO stages, as it also affects shuffle-only stages of the queries. We actually analysed these data across all queries to try to see any correlations between the type of stage and the speedup/slow down, but saw nothing. I am sure the root cause will be something silly...

**Comment by ulysses-you** [comment]
@ted-jenks can you put all spark configs with/without gluten into description ? One idea is, if you enable gluten, then you do not need such big onheap memory, it may cause your executor instance less than vanilla Spark due to the memory resource limitation. In general, you can set 30GB offheap and 6GB onheap with Gluten and 30GB onheap, 6GB overhead without gluten.

**Comment by smlHao** [comment]
> I'm not sure how it performs in a k8s environment, but in general cloud environments are worse. My own test results proved that at least 70% of SQL achieved performance gains. It's not unusual for a few SQLs to be slower than spark, but your test results are still very different from mine.

hi, Don't recommended for use in k8s?

**Comment by ted-jenks** [comment]
As you suggested, I tried runs with 6g onheap and 30g off heap for gluten and the opposite for vanilla spark and unfortunately saw no improvements. 

Config here @ulysses-you:



spark.app.id | spark-e851a77ce1eb4a6e8d6d02e5d1995092
-- | --
spark.app.initial.file.urls | spark://spark-22119a8b28fa029f-driver-svc.compute-spark-2.svc:7078/files/ca.cer,spark://spark-22119a8b28fa029f-driver-svc.compute-spark-2.svc:7078/files/metrics.properties,spark://spark-22119a8b28fa029f-driver-svc.compute-spark-2.svc:7078/files/log4j.xml,spark://spark-22119a8b28fa029f-driver-svc.compute-spark-2.svc:7078/files/truststore.jks
spark.app.name | java.10e9f35f-1fa6-4c07-a935-6df283e18a8b
spark.app.startTime | 1697199563884
spark.authenticate | TRUE
spark.blockManager.port | 7079
spark.decommission.enabled | FALSE
spark.driver.bindAddress | 10.0.82.240
spark.driver.blockManager.port | 7079
spark.driver.cores | 1
spark.driver.extraJavaOptions | -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -Dlog4j.configurationFile=log4j.xml -Dlog4j2.formatMsgNoLookups=true -Dlog4j2.garbagefreeThreadContextMap=true -Dlog4j2.disableJmx=true -Dlog4j2.debug --illegal-access=permit -Dio.netty.tryReflectionSetAccessible=true --add-opens java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED -XX:+CrashOnOutOfMemoryError -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -XX:+UnlockDiagnosticVMOptions -XX:UseAVX=2 -XX:ActiveProcessorCount=1
spark.driver.maxResultSize | 8g
spark.driver.memory | 13g
spark.driver.port | 7078
spark.dynamicAllocation.enabled | FALSE
spark.dynamicAllocation.maxExecutors | 16
spark.dynamicAllocation.shuffleTracking.enabled | TRUE
spark.eventLog.dir | /tmp/spark-events
spark.eventLog.enabled | TRUE
spark.executor.cores | 2
spark.executor.extraJavaOptions | -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -Dlog4j.configurationFile=log4j.xml -Dlog4j2.formatMsgNoLookups=true -Dlog4j2.garbagefreeThreadContextMap=true -Dlog4j2.disableJmx=true -Dlog4j2.debug --illegal-access=permit -Dio.netty.tryReflectionSetAccessible=true --add-opens java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED -XX:+CrashOnOutOfMemoryError -XX:+IgnoreUnrecognizedVMOptions --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.lang.reflect=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/sun.nio.cs=ALL-UNNAMED --add-opens=java.base/sun.security.action=ALL-UNNAMED --add-opens=java.base/sun.util.calendar=ALL-UNNAMED --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED -XX:+UnlockDiagnosticVMOptions -XX:UseAVX=2 -XX:ActiveProcessorCount=1
spark.executor.id | driver
spark.executor.instances | 16
spark.executor.memory | 27g
spark.executor.memoryOverhead | 3g
spark.gluten.loadLibFromJar | TRUE
spark.gluten.memory.conservative.offHeap.size.in.bytes | 16106127360
spark.gluten.memory.conservative.task.offHeap.size.in.bytes | 8053063680
spark.gluten.memory.offHeap.size.in.bytes | 32212254720
spark.gluten.memory.task.offHeap.size.in.bytes | 16106127360
spark.gluten.sql.columnar.backend.lib | velox
spark.gluten.sql.columnar.shuffle.writeEOS | FALSE
spark.gluten.sql.session.timeZone.default | GMT
spark.hadoop.fs.s3a.fast.upload | TRUE
spark.hadoop.fs.s3a.multipart.threshold | 134217728
spark.hadoop.hadoop.tmp.dir | /var/data/spark-local-dir/hadoop-tmp-dir
spark.jars |  
spark.kryoserializer.buffer.max | 2047m
spark.kubernetes.allocation.batch.size | 50
spark.kubernetes.driver.annotation.spark-driver-cpu | 1
spark.kubernetes.driver.annotation.spark-driver-mem | 14643Mi
spark.kubernetes.driver.annotation.spark-driver-nvidia.com/gpu | 0
spark.kubernetes.driver.annotation.spark-executor-count | 16
spark.kubernetes.driver.annotation.spark-executor-cpu | 1.2
spark.kubernetes.driver.annotation.spark-executor-mem | 30720Mi
spark.kubernetes.driver.annotation.spark-executor-nvidia.com/gpu | 0
spark.kubernetes.driver.label.distributed-application-name | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.driver.label.is-driver-pod | TRUE
spark.kubernetes.driver.label.spark-app-id | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.driver.label.spark-backend-id | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.driver.pod.name | java-10e9f35f-1fa6-4c07-a935-6df283e18a8b-a01a768b28fa0282-driver
spark.kubernetes.driver.podTemplateContainerName | spark-container
spark.kubernetes.driver.service.annotation.distributed-application-name | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.driver.service.annotation.distributed-application-resource | TRUE
spark.kubernetes.executor.disableConfigMap | TRUE
spark.kubernetes.executor.label.distributed-application-name | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.executor.label.spark-app-id | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.executor.label.spark-backend-id | 10e9f35f-1fa6-4c07-a935-6df283e18a8b-6994
spark.kubernetes.executor.podNamePrefix | java-10e9f35f-1fa6-4c07-a935-6df283e18a8b-0778a68b28fa30c8
spark.kubernetes.executor.podTemplateContainerName | spark-container
spark.kubernetes.executor.podTemplateFile | /opt/spark/pod-template/pod-spec-template.yml
spark.kubernetes.executor.request.cores | 1.2
spark.kubernetes.resource.type | java
spark.kubernetes.submission.connectionTimeout | 10000
spark.kubernetes.submission.requestTimeout | 300000
spark.kubernetes.submission.waitAppCompletion | FALSE
spark.kubernetes.submitInDriver | TRUE
spark.local.dir | /var/data/spark-local-dir
spark.logConf | TRUE
spark.master | k8s://https://kubernetes.default:443
spark.memory.offHeap.enabled | TRUE
spark.memory.offHeap.size | 30g
spark.metrics.conf | metrics.properties
spark.network.crypto.enabled | TRUE
spark.network.timeout | 300
spark.plugins | io.glutenproject.GlutenPlugin
spark.rdd.compress | TRUE
spark.repl.local.jars |  
spark.rpc.message.maxSize | 512
spark.scheduler.maxRetainedRemovedDecommissionExecutors | 1000
spark.scheduler.maxRetainedUnknownDecommissionExecutors | 1000
spark.scheduler.minRegisteredResourcesRatio | 0
spark.scheduler.mode | FAIR
spark.serializer | org.apache.spark.serializer.KryoSerializer
spark.shuffle.manager | org.apache.spark.shuffle.sort.ColumnarShuffleManager
spark.shuffle.service.enabled | FALSE
spark.smm.keyversion | 3
spark.sql.adaptive.enabled | TRUE
spark.sql.broadcastTimeout | -1
spark.sql.cache.serializer | org.apache.spark.sql.execution.ColumnarCachedBatchSerializer
spark.sql.extensions | io.glutenproject.GlutenSessionExtensions
spark.sql.parquet.int96AsTimestamp | FALSE
spark.stage.ignoreDecommissionFetchFailure | TRUE
spark.storage.decommission.enabled | TRUE
spark.storage.decommission.maxReplicationFailuresPerBlock | 32
spark.storage.decommission.rddBlocks.enabled | TRUE
spark.storage.decommission.replicationReattemptInterval | 10s
spark.storage.decommission.shuffleBlocks.enabled | TRUE
spark.submit.deployMode | client
spark.submit.pyFiles |  
spark.task.reaper.enabled | TRUE
spark.task.reaper.killTimeout | 300s
spark.task.reaper.pollingInterval | 30s
spark.ui.enabled | FALSE

**Comment by ulysses-you** [comment]
@ted-jenks according to the configs, it seems you did not modifiy the onheap memory `spark.executor.memory` for gluten. can you check that ?

**Comment by ted-jenks** [comment]
@ulysses-you Those configs are from another run. I'll do a rerun to get the configs with the lower on-heap memory for you.

**Comment by rshkv** [comment]
@ted-jenks and I noticed that we have large numbers of executor losses when running with Gluten. Jobs do complete in the end (just slower because they run with fewer executors). Failures seem non-deterministic and it's not always as bad as in the run below.

The error we see is the same error reported in #3209. 

```
INFO  [2023-10-17T20:32:49.015979Z] io.glutenproject.vectorized.JniLibLoader: Trying to load library libgluten.so (0: libgluten.so)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00007ff8c8434b37, pid=14, tid=41
#
# JRE version: OpenJDK Runtime Environment Corretto-17.0.8.8.1 (17.0.8.1+8) (build 17.0.8.1+8-LTS)
# Java VM: OpenJDK 64-Bit Server VM Corretto-17.0.8.8.1 (17.0.8.1+8-LTS, mixed mode, tiered, compressed oops, compressed class ptrs, serial gc, linux-amd64)
# Problematic frame:
# C  [libgluten.so+0x461b37]  _GLOBAL__sub_I_LocalPartitionWriter.cc+0x47
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /app/work-dir/core.14)
#
# An error report file with more information is saved as:
# /app/work-dir/hs_err_pid14.log
```


![Web capture_17-10-2023_21424_localhost](https://github.com/oap-project/gluten/assets/7870972/19f1e48a-0049-4542-96e7-5032a34fa1e4)

**Comment by ted-jenks** [comment]
Note: my earlier comment of:
> We are not seeing anything like executor loss or higher stage retry rates in the gluten runs. In the slower queries, stages just run slower than the equivalent stages in vanilla spark. 
Was based on a bad UI. We just found these executor losses and logs today.

**Comment by rshkv** [comment]
The `SIGILL` failing our executors was caused by a `vmovdqa64` (AVX-512F) instruction in `_GLOBAL__sub_I_LocalPartitionWriter`. It turns out that our infra uses a blend of AWS instance types for better availability. And some of those instance types are older and don't support AVX-512.

We've since compiled Gluten without AVX-512 (just AVX) and have no executor failures since.

That said, Gluten is still performing worse than vanilla Spark.

**Comment by ted-jenks** [comment]
For future people encountering the `SIGILL`, we found compiling for Haswell was a good approach. This still gives AVX-256 instruction sets so you benefit from SIMD instructions. Velox itself doesn't use AVX-512 anyway so this shouldn't bring a significant performance hit. See compilation with flag in this [pr](https://github.com/ted-jenks/gluten/pull/4).

**Comment by ted-jenks** [comment]
@rshkv and I have successfully got performance improvements over vanilla Spark with Gluten. Thank you very much @ulysses-you and @kelvin-qin for your help getting set up.

For us the important factors were:
 - Configuring the offheap/onheap memory in line with @ulysses-you's recommendations. 
 - Preventing executor loss due to kubernetes hosts having different instruction set support, but still getting good SIMD perf as per @kelvin-qin's comment by compiling for a target supported by all our hosts (AVX-256 for us).
 - Gluten was emitting far more logs than vanilla Spark. Our production telemetry setup meant this was causing significant overheads. Disabling this gave significant speedup.
 -  We saw compression overheads also being large (especially for shuffle files when compared to vanilla Spark). Turning compression off gave us a boost as we are not network bound.

![better_queries_500](https://github.com/oap-project/gluten/assets/77486246/939ed0ad-3de2-481c-bebc-6a83e9ca42be)
![benchmark_result_2000](https://github.com/oap-project/gluten/assets/77486246/e01eb398-7051-40cc-9978-51cc4370712b)

Thanks again for all the support on this! We have ideas on improving results for our setup further and are super excited about this project.
